### DISCUSSION TITLE
Build failing with 500 error (getStaticProps + Apollo + CORS)

### DISCUSSION THREAD

**Discussion by alex-r89** [vercel/next.js#14731] [Help]
Hi there,

I have an apollo web application which runs a query inside `getStaticProps`. This is working fine in development, and builds fine locally, however fails when building on Vercel. The build console log shows a 500 error.

My assumption is that the build is failing because of the CORS setting on my backend; I allow the origins `https://www.mydomain.com` and `https://mydomain.com`. My API is at `https://api.mydomain.com`.

I assume the build's url is under some vercel domain and therefore the query at build time is being blocked by my CORS policy, however I'm not sure what this url is.

Is there a specific URL I need to add here? Or is there either a way for the build to make requests under my own domain?

Thanks,

**Comment by rafaelalmeidatk** [comment]
`getStaticProps` is called on Node.js side so there is no CORS. Try to log the error message, it will be more helpful

**Reply by alex-r89** [reply]
Sorry if i'm misunderstanding, but the request to my api endpoint must be coming from a Vercel domain/server during the build?

I am logging out CORS rejections on my apollo-express server, and can see a log in the console every time I run a build which made me think it was CORS related.
To explain that, I have this set up in my express server:

```js
  const corsOptions = {
    origin:
      process.env.NODE_ENV === 'production'
        ? (origin: string, callback: any) => {
            if (PROD_LIST.indexOf(origin) !== -1) {
              callback(null, true)
            } else {
              callback(new Error(`${origin} - not allowed by CORS`))
            }
          }
        : 'http://localhost:3000',
    credentials: true
  }

  app.use(cors(corsOptions))
```

Shortly after kicking off a build in Vercel, the above code fires and logs out the following:
```Error: undefined - not allowed by CORS```

I am also logging out the erros in the build process on vercel:

```
17:08:15.396  
SyntaxError [ServerParseError]: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at /vercel/11a0a2d6/node_modules/@apollo/client/apollo-client.cjs.js:1081:25
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  name: 'ServerParseError',
  response: Response {
    size: 0,
    timeout: 0,
    [Symbol(Body internals)]: { body: [PassThrough], disturbed: true, error: null },
    [Symbol(Response internals)]: {
      url: 'https://api.mydomain.com/',
      status: 500,
      statusText: 'Internal Server Error',
      headers: [Headers],
      counter: 0
    }
  },
  statusCode: 500,
  bodyText: '<!DOCTYPE html>\n' +
    '<html lang="en">\n' +
    '<head>\n' +
    '<meta charset="utf-8">\n' +
    '<title>Error</title>\n' +
    '</head>\n' +
    '<body>\n' +
    '<pre>Internal Server Error</pre>\n' +
    '</body>\n' +
    '</html>\n'
}
```

This console log showing a 500, coupled with the CORS message firing on my apollo-express server during the build made me assume its related to CORS, but I guess not?

**Reply by dansteryoo** [reply]
this worked for me by adding the origin in the enhancedFetch function.

const enhancedFetch = (url: any, init: any) => {
  return fetch(url, {
    ...init,
    headers: {
      ...init.headers,
      origin: 'your domain',
      'Access-Control-Allow-Origin': '*',
    },
  }).then((response) => response);
};

const httpLink = new HttpLink({
  uri: 'your domain',
  credentials: 'include',
  fetchOptions: {
    mode: 'cors',
  },
  fetch: enhancedFetch,
});

**Comment by alex-r89** [comment]
I've managed to get this working, however I am unsure if the method I am using is correct/a good idea or if it means I have something configured incorrectly/insecurly:

I simply added an origin header to my apollo HttpLink:

```js
const httpLink = new HttpLink({
  uri: process.env.NODE_ENV === 'production' ? 'https://api.mydomain.com' : 'http://localhost:4000',
  credentials: 'include',
  headers: {
    Origin: 'https://mydomain.com' // <- Added this and now builds are no longer 500 erroring on vercel
  } 
})
```

My getStaticProps method is unchanged:

```js
export async function getStaticProps() {
  const apolloClient = await initApolloClient({})

  await apolloClient.query({
    query: getPhotosQuery,
    variables: { input: { page: 1, getFeaturedPhotosOnly: false } }
  })
  await apolloClient.query({
    query: getPhotosQuery,
    variables: { input: { page: 1, getFeaturedPhotosOnly: true } }
  })

  const apolloStaticCache = apolloClient.cache.extract()
  return {
    props: {
      apolloStaticCache
    },
    unstable_revalidate: 1
  }
}
```

Now, the `getStaticProps`method above doesn't fail with a 500 error when building. However It seems too simple and the fact that I couldn't find this anywhere makes me feel like this solution is wrong?

I was wondering if anyone could advise -  Possibly @Timer, @timneutkens or @lfades who have a better understanding? Apologies for tagging you directly.
